pilopt/src/lib.rs

```diff
@@ -22,16 +23,20 @@ use referenced_symbols::{ReferencedSymbols, SymbolReference};
 
 pub fn optimize<T: FieldElement>(mut pil_file: Analyzed<T>) -> Analyzed<T> {
     let col_count_pre = (pil_file.commitment_count(), pil_file.constant_count());
+    let mut pil_hash = 0;
     remove_unreferenced_definitions(&mut pil_file);
-    remove_constant_fixed_columns(&mut pil_file);
-    deduplicate_fixed_columns(&mut pil_file);
-    simplify_identities(&mut pil_file);
-    extract_constant_lookups(&mut pil_file);
-    remove_constant_witness_columns(&mut pil_file);
-    simplify_identities(&mut pil_file);
-    remove_trivial_identities(&mut pil_file);
-    remove_duplicate_identities(&mut pil_file);
-    remove_unreferenced_definitions(&mut pil_file);
+    while pil_hash != hash_pil_state(&pil_file) {
+        pil_hash = hash_pil_state(&pil_file);
+        remove_constant_fixed_columns(&mut pil_file);
+        deduplicate_fixed_columns(&mut pil_file);
+        simplify_identities(&mut pil_file);
+        extract_constant_lookups(&mut pil_file);
+        remove_constant_witness_columns(&mut pil_file);
+        simplify_identities(&mut pil_file);
+        remove_trivial_identities(&mut pil_file);
+        remove_duplicate_identities(&mut pil_file);
+        remove_unreferenced_definitions(&mut pil_file);
```

Comment:
The main idea behind making this first step out of the loop, was to remove everything that is trivially removable before even calculating the hash, because it will ultimately iterate over definitions, identities, etc.